thunder/dynamo/utils.py

```diff
@@ -555,3 +616,179 @@ def remove_empty_autocast(graph_module: torch.fx.GraphModule) -> torch.fx.GraphM
+        include_device=include_device,
+        colored=colored,
+    )
+    module_code = verbose_python_code.src
+    module_code = module_code.lstrip("\n")
+    module_code = f"class {module_name}(torch.nn.Module):\n" + module_code
+    module_code = _addindent(module_code, 2)
+
+    submodule_code_list = [""]
+    for submodule_name, submodule in module.named_children():
+        if hasattr(submodule, "graph"):
+            submodule_code_list.append(
+                _readable(
+                    submodule,
+                    submodule_name,
+                    print_output=False,
+                    include_stride=include_stride,
+                    include_device=include_device,
+                    colored=colored,
+                )
+            )
+    submodule_code = "\n".join(submodule_code_list)
+    submodule_code = _addindent(submodule_code, 2)
+
+    output = module_code + submodule_code
+    if print_output:
+        print(module_code + submodule_code)
+    return output
+
+
+def get_env() -> tuple[str, str]:
+    """Retrieve detailed environment information using `torch.utils.collect_env.get_pretty_env_info()`.
+    Additionally, include the installed versions of Thunder and NvFuser (if available via pip).
+    """
+
+    from torch.utils.collect_env import run, get_pretty_env_info, get_pip_packages
+
+    torch_env = get_pretty_env_info()
+    _, thunder_packages = get_pip_packages(run, {"lightning-thunder", "nvfuser"})
+    return torch_env, thunder_packages
+
+
+def thunder_options_to_str(thunder_options: dict) -> str:
+    from thunder import resolve_executors
+
+    option_str = ""
+    for key, value in thunder_options.items():
+        if key == "executors":
+            executors = resolve_executors(value)
+            option_str += f"{key}=[" + ",".join(f"thunder.extend.get_executor('{ex.name}')" for ex in executors) + "]"
+        else:
+            option_str += f"{key}={repr(value)}"
+        option_str += ","
+    return option_str
+
+
+def reproducer(
+    gm: torch.fx.GraphModule,
+    thunder_options: dict,
+    args: tuple[torch.Tensor | ExampleInputMetaData],
+    folder: str | os.PathLike,
+    graph_name: str,
+    use_pytest_benchmark: bool = False,
+):
+    folder = Path(folder)
+    folder.mkdir(exist_ok=True)
+    torch_env, thunder_pkgs = get_env()
+    # Ideally we'd use print_readable, but we want verbose=False and there's no
+    # way to set that with print_readable.
+    readable = _readable(gm, "DynamoModule", print_output=False)
+    has_cuda_args = any(hasattr(arg, "device") and arg.device.type == "cuda" for arg in args)
+    thunder_options_str = thunder_options_to_str(thunder_options)
+    with open(folder / f"{graph_name}.py", "w") as f:
+        comment_str = f'''"""
+Environment information get from `torch.utils.collect_env.get_pretty_env_info()`:
+{torch_env}
+
+Versions of Thunder related libraries:
+{thunder_pkgs}
+
+The torch.fx.Graph:
+{gm.graph}
```

Comment:
Why is it useful to save the graph if we're saving the Python code representation for the same graph?

Reply:
I just noticed today there's another Jira for me to save the information about what the original graph module and the split module look like (which part is run by Thunder, which by inductor, and the reason why it's split), I'll create an issue and follow up there